networkx/algorithms/dag.py

```diff
@@ -79,172 +82,183 @@ def is_directed_acyclic_graph(G):
     ----------
     G : NetworkX digraph
+        A directed graph
 
-    nbunch : container of nodes (optional)
-        Explore graph in specified order given in nbunch
+    key : function, optional
+        This function maps nodes to keys with which to resolve ambiguities in
+        the sort order.  Defaults to the identity function.
 
-    reverse : bool, optional
-        Return postorder instead of preorder if True.
-        Reverse mode is a bit more efficient.
+    Returns
+    -------
+    lexicographically_topologically_sorted_nodes : iterable
+        An iterable of node names in lexicographical topological sort order.
 
     Raises
     ------
     NetworkXError
-        Topological sort is defined for directed graphs only. If the
-        graph G is undirected, a NetworkXError is raised.
+        Topological sort is defined for directed graphs only. If the graph G
+        is undirected, a NetworkXError is raised.
 
     NetworkXUnfeasible
-        If G is not a directed acyclic graph (DAG) no topological sort
-        exists and a NetworkXUnfeasible exception is raised.
+        If G is not a directed acyclic graph (DAG) no topological sort exists
+        and a NetworkXUnfeasible exception is raised.  This can also be
+        raised if G is changed while the returned iterator is being processed.
+
+    RuntimeError
+        If G is changed while the returned iterator is being processed.
 
     Notes
     -----
-    This is a recursive version of topological sort.
+    This algorithm is based on a description and proof in
+    Introduction to algorithms - a creative approach [1]_ .
 
     See also
     --------
     topological_sort
-    is_directed_acyclic_graph
 
+    References
+    ----------
+    .. [1] Manber, U. (1989). Introduction to algorithms - a creative approach. Addison-Wesley.
+        http://www.amazon.com/Introduction-Algorithms-A-Creative-Approach/dp/0201120372
     """
     if not G.is_directed():
         raise nx.NetworkXError(
             "Topological sort not defined on undirected graphs.")
 
-    def _dfs(v):
-        ancestors.add(v)
+    if key is None:
+        key = lambda x: x
 
-        for w in G[v]:
-            if w in ancestors:
-                raise nx.NetworkXUnfeasible("Graph contains a cycle.")
+    def create_tuple(node):
+        return key(node), node
 
-            if w not in explored:
-                _dfs(w)
+    indegree_map = {v: d for v, d in G.in_degree() if d > 0}
+    # These nodes have zero indegree and ready to be returned.
+    zero_indegree = [create_tuple(v) for v, d in G.in_degree() if d == 0]
+    heapq.heapify(zero_indegree)
 
-        ancestors.remove(v)
-        explored.add(v)
-        order.append(v)
+    while zero_indegree:
+        _, node = heapq.heappop(zero_indegree)
 
-    ancestors = set()
-    explored = set()
-    order = []
+        if node not in G:
+            raise RuntimeError("Graph changed during iteration")
+        for _, child in G.edges(node):
```

Comment:
How about some simple tests?  If there is more than one valid result allow
any.
On Aug 7, 2015 7:44 PM, "Dan Schult" <notifications@github.com> wrote:

> In networkx/algorithms/dag.py
> <https://github.com/networkx/networkx/pull/1662#discussion_r36573497>:
>
> >
> > -    ancestors = set()
> > -    explored = set()
> > -    order = []
> > +        if node not in G:
> > +            raise RuntimeError("Graph changed during iteration")
> > +        for _, child in G.edges(node):
>
> G[node] is a direct lookup of the successors of node. G.edges(node)
> returns outgoing edges that involve node. It wraps the successors into a
> tuple with the node. Sometimes the tuple is helpful, but here I think you
> just unwrap it again to get the successor. Not a big deal, but G[node] is
> typically what you want if you already know node.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/networkx/networkx/pull/1662/files#r36573497>.
>